src/test/webServer.ts

```diff
@@ -0,0 +1,140 @@
+};
+
+const newProcessLogPrefixer = () => new stream.Transform({
+  transform(this: stream.Transform, chunk: Buffer, encoding: string, callback: stream.TransformCallback) {
+    this.push(chunk.toString().split(os.EOL).map((line: string): string => line ? `[WebServer] ${line}` : line).join(os.EOL));
+    callback();
+  },
+});
+
+export class WebServer {
+  private _killProcess?: () => Promise<void>;
+  private _processExitedWithNonZeroStatusCode!: Promise<any>;
+  constructor(private readonly config: WebServerConfig) { }
+
+  public static async create(config: WebServerConfig): Promise<WebServer> {
+    const webServer = new WebServer(config);
+    if (config.port)
+      await webServer._verifyFreePort(config.port);
+    try {
+      const port = await webServer._startWebServer();
+      await webServer._waitForAvailability(port);
+      process.env.PLAYWRIGHT_TEST_BASE_URL = `http://localhost:${port}`;
+      return webServer;
+    } catch (error) {
+      await webServer.kill();
+      throw error;
+    }
+  }
+
+  private async _verifyFreePort(port: number) {
+    const cancellationToken = { canceled: false };
+    const portIsUsed = await Promise.race([
+      new Promise(resolve => setTimeout(() => resolve(false), 100)),
+      waitForSocket(port, 100, cancellationToken),
+    ]);
+    cancellationToken.canceled = true;
+    if (portIsUsed)
+      throw new Error(`Port ${port} is used, make sure that nothing is running on the port`);
+  }
+
+  private async _startWebServer(): Promise<number> {
+    let collectPortResolve = (port: number) => { };
+    const collectPortPromise = new Promise<number>(resolve => collectPortResolve = resolve);
+    function collectPort(data: Buffer) {
+      const regExp = /http:\/\/localhost:(\d+)/.exec(data.toString());
+      if (regExp)
+        collectPortResolve(parseInt(regExp[1], 10));
+    }
+
+    const { launchedProcess, kill } = await launchProcess({
+      command: this.config.command,
+      env: {
+        ...DEFAULT_ENVIRONMENT_VARIABLES,
+        ...process.env,
+        ...this.config.env,
+      },
+      cwd: this.config.cwd,
+      stdio: 'pipe',
+      shell: true,
+    });
+    this._killProcess = kill;
+
+    launchedProcess.stdout.pipe(newProcessLogPrefixer()).pipe(process.stdout);
+    launchedProcess.stderr.pipe(newProcessLogPrefixer()).pipe(process.stderr);
+    if (!this.config.port)
+      launchedProcess.stdout.on('data', collectPort);
+    let processExitedWithNonZeroStatusCodeCallback = (error: Error) => { };
+    this._processExitedWithNonZeroStatusCode = new Promise((_, reject) => processExitedWithNonZeroStatusCodeCallback = reject);
+    launchedProcess.on('exit', code => processExitedWithNonZeroStatusCodeCallback(new Error(`WebServer was not able to start. Exit code: ${code}`)));
+    if (this.config.port)
+      return this.config.port;
+    console.log(`Starting WebServer port detection.`);
```

Comment:
I'd move this to the start of the function and change it to 'Starting web server with ${command}...'